lxd/api_metrics.go

```diff
@@ -331,31 +331,30 @@ func metricsGet(d *Daemon, r *http.Request) response.Response {
 }
 
 func getFilteredMetrics(s *state.State, r *http.Request, compress bool, metricSet *metrics.MetricSet) response.Response {
+	// Ignore filtering in case the authentication for metrics is disabled.
 	if !s.GlobalConfig.MetricsAuthentication() {
 		return response.SyncResponsePlain(true, compress, metricSet.String())
 	}
 
-	// Get instances the user is allowed to view.
-	userHasPermission, err := s.Authorizer.GetPermissionChecker(r.Context(), r, auth.EntitlementCanView, entity.TypeInstance)
-	if err != nil && !auth.IsDeniedError(err) {
+	userHasProjectPermission, err := s.Authorizer.GetPermissionChecker(r.Context(), r, auth.EntitlementCanViewMetrics, entity.TypeProject)
+	if err != nil {
 		return response.SmartError(err)
-	} else if err != nil {
-		// This is counterintuitive. We are unauthorized to get a permission checker for viewing instances because a metric type certificate
-		// can't view instances. However, in order to get to this point we must already have auth.EntitlementCanViewMetrics. So we can view
-		// the metrics but we can't do any filtering, so just return the metrics.
-		return response.SyncResponsePlain(true, compress, metricSet.String())
 	}
 
-	// Filter by project name and instance name.
-	metricSet.FilterSamples(userHasPermission)
-
-	userHasPermission, err = s.Authorizer.GetPermissionChecker(r.Context(), r, auth.EntitlementCanView, entity.TypeProject)
+	userHasServerPermission, err := s.Authorizer.GetPermissionChecker(r.Context(), r, auth.EntitlementCanViewMetrics, entity.TypeServer)
 	if err != nil {
 		return response.SmartError(err)
 	}
 
-	// Filter by project only.
-	metricSet.FilterSamples(userHasPermission)
+	// Filter all metrics for view permissions.
+	// Internal server metrics without labels are compared against the server entity.
+	metricSet.FilterSamples(func(labels map[string]string) bool {
+		if labels["project"] != "" {
```

Comment:
Is this correct? What if the label does have a non-empty project, but the user doesn't have a project permission, but does have server permission, this would return false?

Doesn't having the server permission mean they can see all metrics (including the project ones)?

Reply:
Okay tricky one, thanks for rising this as I have found another potential security bug in my changes. The unrestricted metrics certificate got just too much permission.

But the way the checking is done here is right.
In case of the TLS driver if the certificate is unrestricted, the driver permits `can_view_metrics` by default (see latest push where I went back to the original changes). 
In case of OIDC this is allowed as the `can_view_metrics` on the `server` type supersedes the `can_view_metrics` on the `project` type. This way the check still returns `true` even if there is no explicit `can_view_metrics` entitlement set for the project (see the changed OpenFGA model).

Reply:
OK so its resolved internally in each driver that the user has implicit project level permissions if its assigned at the server level. Makes sense thanks